hugo/content/docs/sematic-model.md

```diff
@@ -319,25 +327,44 @@ X:
     {A} 'A' name=ID 
   | {B} 'B' name=ID count=INT;
 ```
+
 Note the absence of the keyword `infer` compared to [actions which infer a type](#simple-actions).
 
-## Refactoring Dummy Rules
-*Dummy rules* are parser rules that are not reachable from the entry rule of the grammar. Despite the fact that they do not participate in the parsing process, they still influence the shape of the semantic model. They infer types in the same fashion as any other parser rule with type inference. However, because dummy rules are exempt from validation checks they are prone to introduce breaking changes. For this reason, we strongly advise using declared types instead of dummy rules.
+## Reference Unions
+
+Trying to reference different types of elements can be an error prone process. Take a look at the following rule which tries to reference either a `Function` or a `Variable`:
 
-Let's look at two dummy rules:
 ```
-X : A | B;
+MemberCall: (element=[Function:ID] | element=[Variable:ID]);
+```
+
+As both alternatives are only an `ID` from a parser perspective, this grammar is not decidable and the `langium` CLI script will throw an error during generation. Luckily, we can improve on this by adding a layer of indirection using an additional parser rule:
 
-Y: name=ID;
 ```
-The first one is inferring a type alias, and the second one an interface. They are similar to parser rules explained in the [previous section](#inferred-types), but they are not reachable from the entry rule of the grammar. They should be replaced with **declared types** as follows:
+NamedElement: Function | Variable;
+
+MemberCall: element=[NamedElement:ID];
 ```
-type X = A | B;
 
-interface Y {
+This allows us to reference either `Function` or `Variable` using the common rule `NamedElement`. However, we have now introduced a rule which is never actually parsed, but only exists for the purpose of the type system to pick up on the correct target types of the reference. Using declared types, we are able to refactor this unused rule, making our grammar more resilient in the process:
+
+```
+// Note the `type` prefix here
+type NamedElement = Function | Variable;
+
+MemberCall: element=[NamedElement:ID];
+```
+
+We can also use interfaces in place of union types with similar results:
+
+```
+interface NamedElement {
     name: string
 }
 
-Y returns Y: name=ID;
+// Infers an interface `Function` that extends `NamedElement`
```

Comment:
I'd suggest to tell about the difference in using `NamedElement` as an interface and as a type. Specifically, that different types will be generated and there will be no `Function` type if you use the syntax `Function returns NamedElement`.